### PR TITLE
Disable FlashAttenion for is_causal=True when seqlen q not equal kv

### PR DESCRIPTION
**Pull request by drisspg**
# Summary:
This pull request **removes** support for non-square sequence lengths in causal attention when using FlashAttention V2.

### Why are doing this
  // FlashAttention 2 updated the default mask meaning for causal in this PR:
  // 9e5e8bc91e it is now aligned to lower_right which would be a BC break
  // for non-square masks. We will not support non-square masks for causal w/ FAV2
  
 For more context see:
 https://github.com/pytorch/pytorch/issues/108108
 
 ### Followup
 A large number of people will likely want to use FAV2 with lower_right causal attention for non equal sequence lengths. See this RFC : https://github.com/pytorch/pytorch/issues/110681

### CODE REVIEW

File: aten/src/ATen/native/transformers/cuda/sdp_utils.cpp [hunk 1]
```diff
@@ -230,6 +230,23 @@ bool check_requires_grad_and_head_dim_gt192_and_sm_ge86_lt90(
   return true;
 }
 
+
+bool check_flash_causal_non_square_seqlens(sdp_params const& params, bool debug) {
+  // FlashAttention 2 updated the default mask meaning for causal in this PR:
+  // 9e5e8bc91e it is now aligned to lower_right which would be a BC break
+  // for non-square masks. We will not support non-square masks for causal w/ FAV2
+  if (params.is_causal && params.query.size(-2) != params.key.size(-2)) {
```
Comment (drisspg) [inline aten/src/ATen/native/transformers/cuda/sdp_utils.cpp:238]: @ezyang this is failing dynamic shapes test e.g.

`pytest  test/dynamo/test_dynamic_shapes.py -k sdpa`
w/ 

``` Python
torch._dynamo.exc.TorchRuntimeError: Failed running call_function <built-in function scaled_dot_product_attention>(*(FakeTensor(..., device='cuda:0', size=(1, s0, 1, s0), dtype=torch.float16,
           requires_grad=True), FakeTensor(..., device='cuda:0', size=(1, s0, 1, s0), dtype=torch.float16,
           requires_grad=True), FakeTensor(..., device='cuda:0', size=(1, s0, 1, s0), dtype=torch.float16,
           requires_grad=True), None, 0, True), **{}):
Cannot call sizes() on tensor with symbolic sizes/strides
Exception raised from throw_cannot_call_with_symbolic at /home/drisspg/meta/pytorch/c10/core/TensorImpl.cpp:593 (most recent call first):
```

Comment (drisspg) [reply aten/src/ATen/native/transformers/cuda/sdp_utils.cpp:238]: https://hud.pytorch.org/pr/111007

**Review by cpuhrsch — APPROVED** [review]

**Comment by drisspg** [comment]
@Pytorchbot merge -f "unrelated failures"

**Pull request closed without merging**